File: backend/app/schemas/design_flows.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


def empty_graph() -> dict[str, Any]:
    return {"nodes": [], "edges": [], "viewport": {"x": 0, "y": 0, "zoom": 1}}


class DesignFlowCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=2000)
    graph_json: dict[str, Any] | None = None


class DesignFlowUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=2000)
    graph_json: dict[str, Any] | None = None


class DesignFlowRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    description: str | None
    graph_json: dict[str, Any]
    created_by: str | None
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/designFlows.ts
import { z } from "zod";

export const DesignGraphSchema = z.object({
  nodes: z.array(z.any()).default([]),
  edges: z.array(z.any()).default([]),
  viewport: z
    .object({
      x: z.number(),
      y: z.number(),
      zoom: z.number(),
    })
    .default({ x: 0, y: 0, zoom: 1 }),
});
export type DesignGraph = z.infer<typeof DesignGraphSchema>;

export const DesignFlowSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string().nullable(),
  graph_json: DesignGraphSchema,
  created_by: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});
export type DesignFlow = z.infer<typeof DesignFlowSchema>;

export type DesignFlowCreatePayload = {
  name: string;
  description?: string | null;
  graph_json?: DesignGraph;
};

export type DesignFlowUpdatePayload = {
  name?: string;
  description?: string | null;
  graph_json?: DesignGraph;
};

export function emptyDesignGraph(): DesignGraph {
  return { nodes: [], edges: [], viewport: { x: 0, y: 0, zoom: 1 } };
}
